storm-core/src/jvm/backtype/storm/windowing/WindowManager.java

```diff
@@ -0,0 +1,210 @@
+     */
+    public void add(T event) {
+        add(event, System.currentTimeMillis());
+    }
+
+    /**
+     * Add an event into the window, with the given ts as the tracking ts.
+     *
+     * @param event the event to track
+     * @param ts    the timestamp
+     */
+    public void add(T event, long ts) {
+        Event<T> windowEvent = new EventImpl<T>(event, ts);
+        window.add(windowEvent);
+        track(windowEvent);
+        compactWindow();
+    }
+
+    /**
+     * The callback invoked by the trigger policy.
+     */
+    @Override
+    public void onTrigger() {
+        List<Event<T>> windowEvents = new ArrayList<>();
+        List<T> expired = null;
+        try {
+            lock.lock();
+            /*
+             * scan the entire window to handle out of order events in
+             * the case of time based windows.
+             */
+            expireEvents(true, windowEvents);
+            expired = new ArrayList<>(expiredEvents);
+            expiredEvents.clear();
+        } finally {
+            lock.unlock();
+        }
+        List<T> events = new ArrayList<>();
+        List<T> newEvents = new ArrayList<>();
+        for (Event<T> event : windowEvents) {
+            events.add(event.get());
+            if (!prevWindowEvents.contains(event)) {
+                newEvents.add(event.get());
+            }
+        }
+        prevWindowEvents.clear();
+        prevWindowEvents.addAll(windowEvents);
+        LOG.debug("invoking windowLifecycleListener onActivation, [{}] events in window.", windowEvents.size());
+        windowLifecycleListener.onActivation(events, newEvents, expired);
+        triggerPolicy.reset();
+    }
+
+    public void shutdown() {
+        LOG.debug("Shutting down WindowManager");
+        if (triggerPolicy != null) {
+            triggerPolicy.shutdown();
+        }
+    }
+
+    /**
+     * expires events that fall out of the window every
+     * EXPIRE_EVENTS_THRESHOLD so that the window does not grow
+     * too big.
+     */
+    private void compactWindow() {
+        if (eventsSinceLastExpiry.incrementAndGet() >= EXPIRE_EVENTS_THRESHOLD) {
+            expireEvents(false, null);
+        }
+    }
+
+    /**
+     * feed the event to the eviction and trigger policies
+     * for bookkeeping and optionally firing the trigger.
+     */
+    private void track(Event<T> windowEvent) {
+        evictionPolicy.track(windowEvent);
+        triggerPolicy.track(windowEvent);
+    }
+
+    /**
+     * Expire events from the window, using the expiration policy to check
+     * if the event should be evicted or not.
+     *
+     * @param fullScan  if set, will scan the entire window. if not set, will stop
+     *                  as soon as an event not satisfying the expiration policy is found.
+     * @param remaining the list of remaining events in the window after expiry.
+     */
+    private void expireEvents(boolean fullScan, List<Event<T>> remaining) {
```

Comment:
why do we take a remaining list as input and mutate it? Why not just return the remaining list? same thing with expired events? Any reason for mutating instead of just returning?